src/shims/foreign_items/posix.rs

```diff
@@ -23,43 +24,58 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         match link_name {
             // Environment related shims
             "getenv" => {
-                let result = this.getenv(args[0])?;
+                let &[name] = check_args(args)?;
+                let result = this.getenv(name)?;
                 this.write_scalar(result, dest)?;
             }
             "unsetenv" => {
-                let result = this.unsetenv(args[0])?;
+                let &[name] = check_args(args)?;
+                let result = this.unsetenv(name)?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
             "setenv" => {
-                let result = this.setenv(args[0], args[1])?;
+                let &[name, value, _overwrite] = check_args(args)?;
+                let result = this.setenv(name, value)?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
             "getcwd" => {
-                let result = this.getcwd(args[0], args[1])?;
+                let &[buf, size] = check_args(args)?;
+                let result = this.getcwd(buf, size)?;
                 this.write_scalar(result, dest)?;
             }
             "chdir" => {
-                let result = this.chdir(args[0])?;
+                let &[path] = check_args(args)?;
+                let result = this.chdir(path)?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
 
             // File related shims
             "open" | "open64" => {
-                let result = this.open(args[0], args[1])?;
+                let &[path, flag, _mode] = check_args(args)?;
+                let result = this.open(path, flag)?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
             "fcntl" => {
-                let result = this.fcntl(args[0], args[1], args.get(2).cloned())?;
+                let result = if args.len() == 2 {
+                    let &[fd, cmd] = check_args(args)?;
+                    this.fcntl(fd, cmd, None)?
+                } else {
+                    // If args.len() isn't 2 or 3 this will error appropriately.
+                    let &[fd, cmd, start] = check_args(args)?;
+                    this.fcntl(fd, cmd, Some(start))?
+                };
```

Comment:
I think it would be better to just pass `args` to `fcntl`, and let it do the checking.